### ISSUE TITLE
option to open internal links in another map view of same map

### ISSUE THREAD

**Issue by kenfhill84083** [freeplane/freeplane#316]
It would be helpful for me if I could click on a node with an internal link in the same map and have the link open the target node in a 2nd map view I already have open.

I was taking sermon notes in a map that had branches for the sermon notes and another branch for Bible cross references.  I had opened a new map view and brought the cross references into view 2 while working on the notes in view 1.   Several times I caught myself thinking, "I wish my notes in view 1 would stay in place and the view 2 would navigate to the target cross reference."  At the same time, I would like the focus to remain on view 1 where I was working on my notes.

That way, I could keep working seamlessly in view 1 

As it is now, when I clicked the node in view 1, it navigated to the target cross reference also in view 1.  It was kind of dizzying.

![open links in 2nd map view](https://user-images.githubusercontent.com/6380301/163484145-8f7364c8-0058-4d1e-ba17-5fdd6e30509f.jpg)

**Comment by kenfhill84083** [comment]
It seems I am the only user who would like this feature.  I have developed a workaround by (1) in map A, on the node with the link, I follow Edit > Copy > Copy node ID.  Then (2) going to map view B I invoke Navigate > Goto node > and paste in the ID I copied.

Is there any easier workaround?

**Comment by macmarrum** [comment]
@kenfhill84083 This is an interesting idea. I like it :+1: 

Here's a script that can act as an easier workaround: [selectLinkedNodeInOtherView.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/selectLinkedNodeInOtherView.groovy)

You can assign it a hot key (e.g. `Ctrl+Shift+Enter`) and use it on your nodes with links.

Please let me know if this works for you.

**Comment by macmarrum** [comment]
I've renamed the script to [selectLinkedNodeInAnotherView.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/selectLinkedNodeInAnotherView.groovy)

Also improved it to select the same node in another view if no link is found.

The view sync can be disabled in the script, if undesired.

**Comment by macmarrum** [comment]
I should also mention that you can remove `="/menu_bar/Mac1/Select"` from the script's first line to have the script appear in `Tools->Scripts` and in `right-click (on a node) -> Scripts`.

The last one could be a way to use the script with your mouse alone.

**Comment by dpolivaev** [comment]
I can implement the logic from the script as an option so that Freeplane opens links in the other view if it is available.

However, if you use view 1 for navigation and view 2 for selection and click on a link in view 2 it would open the link in view 1, and it is probably not what you want. What you actually want is that all links to be opened on view 2.

I can add a menu item to mark views as views for links to be opened. But I do not know how to explain it to any other user.

So, unless we find a solution with documentation implementing this feature seems to be not the right thing to do because it would just confuse most people.

**Comment by macmarrum** [comment]
> I can implement the logic from the script as an option so that Freeplane opens links in the other view if it is available.

Excellent :+1: 

> However, if you use view 1 for navigation and view 2 for selection and click on a link in view 2 it would open the link in view 1, and it is probably not what you want. What you actually want is that all links to be opened on view 2.

The way I imagine it working is that a regular click will continue the current logic but a click with a modifier key (e.g. with `Shift`) will switch to another view.

This way my first `Shift+click` will take me to another view where I can use regular clicks for moving to linked nodes within the view; and if I want to go "back" to the other view, I again can use `Shift+click`. In this approach no special explanation would be required.

**Comment by dpolivaev** [comment]
How can we communicate it to other users when it is implemented?

**Comment by macmarrum** [comment]
From you comment I'm getting that the usual places where new features are described might be inappropriate or insufficient.

When I think about it, there are more such obscure keyboard + mouse combinations, like `Ctrl+click` to set a hot key, `Ctrl+C` to copy from Command Search or Filter Composer, and many more. I can create a page in the docs to describe them all.

I know that some other applications can display "tip of the day", e.g. LibreOffice or IntelliJ IDEA. Have you ever considered to add it to Freeplane?

**Comment by dpolivaev** [comment]
Done, happy testing from the source

**Comment by kenfhill84083** [comment]
> I've renamed the script to [selectLinkedNodeInAnotherView.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/selectLinkedNodeInAnotherView.groovy)
> 
> Also improved it to select the same node in another view if no link is found.
> 
> The view sync can be disabled in the script, if undesired.

Thanks for your interest and work on this script.  Unfortunately it does not do anything beyond what clicking the link does - navigate to the target node but in the same map view.  I can make a video if you want.

WOW.  Lots of activity while I slept.

**Comment by macmarrum** [comment]
> Done, happy testing from the source

It looks good to me :+1:

**Comment by macmarrum** [comment]
> Thanks for your interest and work on this script. Unfortunately it does not do anything beyond what clicking the link does - navigate to the target node but in the same map view. I can make a video if you want.

Yes, please. A video might help.
Probably more helpful would be the actual mind map (or a portion of it containing the node with a link).

> WOW. Lots of activity while I slept.

:smile:

**Comment by kenfhill84083** [comment]
> Yes, please. A video might help. Probably more helpful would be the actual mind map (or a portion of it containing the node with a link).

[freeplane _ map view script behavior (video - kfh).zip](https://github.com/freeplane/freeplane/files/11781977/freeplane._.map.view.script.behavior.video.-.kfh.zip)
[script test (MM - kfh).zip](https://github.com/freeplane/freeplane/files/11781979/script.test.MM.-.kfh.zip)

**Comment by dpolivaev** [comment]
Preview 1.11.5_11 :  mouse click + shift opens linked node in another view if it is available

**Comment by kenfhill84083** [comment]
> Preview 1.11.5_11 : mouse click + shift opens linked node in another view if it is available

Fantastic!  Thanks.  Your responsiveness continues to boggle my mind as it increases the value of Freeplane.

**Comment by macmarrum** [comment]
@dpolivaev I can see that `mouse click + shift` opens linked node in another view if it is available and in the same view if not. 

I'd expect the link **not** to be opened when another view is missing, and to see a message  (e.g. in `c.statusInfo`) that explains the apparent lack of action.
This would support my flow of switching to another view but stopping if this view is unavailable. Otherwise I need to `Alt+BACK` to last node before `View->New map view`.

In my opinion, the new `Shift+Click` would be easier to explain to others: "it takes you to another view, but only if it exists."
And there would be less room for confusion about `Preferences…->Behaviour->Open node link on simple mouse click`, i.e. `Ctrl+Click` or `(simply) Click` does the regular job of following hyperlinks (within the mind map or opening external mind maps), whereas `Shift+Click` follows hyperlinks to nodes in another view.

@kenfhill84083 @dpolivaev What do you think?

**Comment by kenfhill84083** [comment]
> 

Thinking about the way I work, my choice would be to have FP open a 2nd map view and navigate to the target in map view 2.  But I suspect that is not what most users would expect.

In a big map, I might find it disconcerting to accidentally navigate to a node, possibly at the opposite, far-distant part of the map - then alt+left back to the launching node.  I suppose I'd get used to it.  On the other hand, the way it works now might act - for me at least - as a welcome reminder to open the 2nd map view.  I'm just guessing.

But if nothing happened, I might be equally confused - I am not in the habit of checking the status bar when things seem not to work.  But I should get into that habit.  (Usually it is because command search is open and focused - often, out of mind, on a 2nd monitor.)

Frankly, I am so happy with this new behavior I can go with whatever seems OK with you guys to limit "collateral damage".

**Comment by macmarrum** [comment]
> But if nothing happened, I might be equally confused

That's a good point. This got me thinking that `Shift+Click` could create a new map view if it didn't exist, before select the target node in that view. People might be familiar with this concept from web browsers.

My revised idea
`Click` or `Ctrl+Click` -> regular
`Shift+Click` -> select in another view, creating it first, if not yet available

**Comment by kenfhill84083** [comment]
> My revised idea `Click` or `Ctrl+Click` -> regular `Shift+Click` -> select in another view, creating it first, if not yet available

This proposal would be my choice.

**Comment by kenfhill84083** [comment]
I have one more hope for this functionality.  Would it be possible to shift+click links in details and notes and get the same result as shift+click on the node link - navigating to the target in the other map view?

**Comment by macmarrum** [comment]
Ah, yes. The HTML links, e.g. 
```html
<a href="#ID_123456">A B C</a>
```

**Comment by dpolivaev** [comment]
I am not sure we want to spend shift + mouse click entirely on this purpose. Once it is done we can not use it for any other purpose.

**Comment by dpolivaev** [comment]
There are also other actions where nodes are selected, for example, goto-operations related to connectors and clones.
They all no nothing about the mouse event and the shift key, and I am hesitant to introduce a general coupling letting the part which is responsible for the selection know anything about the mouse event.

The current solution is clean because the mouse event is checked by the mouse listener of the node, but if the link is inside editor the listener is implemented by the editor.

If I do one step to help, you want me to do 1000 more. I don't like it.

**Comment by kenfhill84083** [comment]
> The current solution is clean because the mouse event is checked by the mouse listener of the node, but if the link is inside editor the listener is implemented by the editor.

Got it.  Thanks for the response.

**Comment by macmarrum** [comment]
> If I do one step to help, you want me to do 1000 more. I don't like it.

@dpolivaev  It wasn't my intention to put any pressure on you. Therefore, I am sorry if my actions/words caused any discomfort.

My only desire is to explain my needs and ides and describe them as best I can so that a discussion can take place. I certainly don't expect all my ideas to end up as implementations.

**Comment by dpolivaev** [comment]
I am happy when I find a solution that just is good enough for a given use case. I am not trying to implement anything perfect because this world is not perfect and my resources are limited. I appreciate it if you could always keep it in mind.

**Comment by kenfhill84083** [comment]
>I am not trying to implement anything perfect because this world is not perfect and my resources are limited. I appreciate it if you could always keep it in mind.

Perfectly reasonable.  

I have a tendency to seek solutions that best suit my needs/wants - in hopes they are within the realm of the possible.  Being pig-ignorant of programming constraints, I had no idea about the complex scope and limitations of mouse listeners spread throughout the code.  

My purpose is never to push you into wastefully sacrificing either your time or Freeplane's efficiency on useless pursuits.  Sorry if I came across as demanding - here and elsewhere.

I am very grateful for all you do and for your wonderful, historic accomplishment in Freeplane.  And the lastest enhancement - as is - in this thread has me almost giddy with joy.

Thanks.

**Comment by dpolivaev** [comment]
There is a preview 1.11.5_12 which opens a second map view if it is not there yet.
I would like you to postpone discussing any other ideas before 1.11.5 is released as stable because I do not want to delay the shipment of critical bug fixes because some of the new features are not perfect yet.

Regards.

**Comment by dpolivaev** [comment]
> From you comment I'm getting that the usual places where new features are described might be inappropriate or insufficient.

There are even more helpful features like the command search dialog not discovered by the users.

> When I think about it, there are more such obscure keyboard + mouse combinations, like Ctrl+click to set a hot key, Ctrl+C to copy from Command Search or Filter Composer, and many more. I can create a page in the docs to describe them all.

👍 

> I know that some other applications can display "tip of the day", e.g. LibreOffice or IntelliJ IDEA. Have you ever considered to add it to Freeplane?

Personally, I never like "tips of the day" or similar stuff because they always come at the wrong time. 
Probably, things will change soon when AI gets so good that users can ask it any question about any software like Freeplane and get a correct answer. Because the AI would have the capability to read all of the documentation and maybe even experiment with the software and describe it.

**Comment by macmarrum** [comment]
> There is a preview 1.11.5_12 which opens a second map view if it is not there yet.

Thank you :heart: 
I've tested it: works well for me :ok_hand:

**Comment by macmarrum** [comment]
> When I think about it, there are more such obscure keyboard + mouse combinations, like Ctrl+click to set a hot key, Ctrl+C to copy from Command Search or Filter Composer, and many more. I can create a page in the docs to describe them all.

Here's a new page in docs, with a description about `Shift+click` to open a new map view:

https://docs.freeplane.org/user-documentation/hot-keys-and-beyond.html

Any comments are welcome

**Comment by EdoFro** [comment]
just a comment:
I use the ctrl key to click on links, to avoid involuntary openings. I would like that when having set that option in preferences, then this new feature would use ctrl +shift.

when selecting multiple nodes using shift +click can launch this new feature involuntary 

thanks,
edo

**Comment by EdoFro** [comment]
> My revised idea
> `Click` or `Ctrl+Click` -> regular
> `Shift+Click` **_or  `Shift+ Ctrl+Click`_** -> select in another view, creating it first, if not yet available

**Comment by kenfhill84083** [comment]
> Here's a new page in docs, with a description about `Shift+click` to open a new map view:
> 
> https://docs.freeplane.org/user-documentation/hot-keys-and-beyond.html
> 
> Any comments are welcome

I had never heard about the Method 3 bonus - good to know.

In the section for assigning hotkeys, I think a heads up would be good for when you try to assign an already-assigned key combo - with a screenshot:
![Freeplane _ currently assigned message](https://github.com/freeplane/freeplane/assets/6380301/e43e964a-6d5e-4b0a-80c4-54f24fdbe070)

**Comment by macmarrum** [comment]
> I had never heard about the Method 3 bonus - good to know.

:+1: 

> In the section for assigning hotkeys, I think a heads up would be good for when you try to assign an already-assigned key combo - with a screenshot

I agree.
Added -> https://macmarrum.github.io/freeplane-docs/user-documentation/hot-keys-and-beyond.html#hot-key-reassignment

**Comment by kenfhill84083** [comment]
> Added -> https://macmarrum.github.io/freeplane-docs/user-documentation/hot-keys-and-beyond.html#hot-key-reassignment

Looks good.

**Comment by macmarrum** [comment]
> Thinking about the way I work, my choice would be to have FP open a 2nd map view and navigate to the target in map view 2. But I suspect that is not what most users would expect.
> 
> In a big map, I might find it disconcerting to accidentally navigate to a node, possibly at the opposite, far-distant part of the map - then alt+left back to the launching node. I suppose I'd get used to it. On the other hand, the way it works now might act - for me at least - as a welcome reminder to open the 2nd map view. I'm just guessing.
> 
> But if nothing happened, I might be equally confused - I am not in the habit of checking the status bar when things seem not to work. But I should get into that habit. (Usually it is because command search is open and focused - often, out of mind, on a 2nd monitor.)
> 
> Frankly, I am so happy with this new behavior I can go with whatever seems OK with you guys to limit "collateral damage".

I have extended [selectLinkedNodeInAnotherView.groovy](https://raw.githubusercontent.com/macmarrum/freeplane-scripting/main/src/scripts/selectLinkedNodeInAnotherView.groovy) to open a new map view, if needed. And because the view is created as a new tab, a pop-up is displayed not to miss it (the pop-up can be easily commented out in the script). I then usually move the new view, so that I can see both views side by side.

> > > Thanks for your interest and work on this script. Unfortunately it does not do anything beyond what clicking the link does - navigate to the target node but in the same map view. I can make a video if you want.
> > >
> > Yes, please. A video might help. Probably more helpful would be the actual mind map (or a portion of it containing the node with a link).
> >
> [freeplane _ map view script behavior (video - kfh).zip](https://github.com/freeplane/freeplane/files/11781977/freeplane._.map.view.script.behavior.video.-.kfh.zip)
> [script test (MM - kfh).zip](https://github.com/freeplane/freeplane/files/11781979/script.test.MM.-.kfh.zip)

The issue is fixed.
